org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/workspace/ModelProviderUtil.java

```diff
@@ -9,30 +9,61 @@
 package org.eclipse.buildship.core.internal.workspace;
 
 import java.util.Collection;
-import java.util.Set;
 
+import org.gradle.tooling.BuildAction;
 import org.gradle.tooling.CancellationTokenSource;
+import org.gradle.tooling.ProjectConnection;
+import org.gradle.tooling.model.build.BuildEnvironment;
 import org.gradle.tooling.model.eclipse.EclipseProject;
 
-import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableList;
 
 import org.eclipse.core.runtime.IProgressMonitor;
 
-import org.eclipse.buildship.core.internal.util.gradle.HierarchicalElementUtils;
+import org.eclipse.buildship.core.internal.util.gradle.IdeFriendlyClassLoading;
+import org.eclipse.buildship.core.internal.util.gradle.GradleVersion;
+import org.eclipse.buildship.core.internal.util.gradle.SimpleIntermediateResultHandler;
 
 public class ModelProviderUtil {
 
-    /**
-     * @return a flattened list of Eclipse projects for all projects included in the Gradle build
-     */
-    public static Set<EclipseProject> fetchAllEclipseProjects(InternalGradleBuild build, CancellationTokenSource tokenSource, FetchStrategy fetchStrategy, IProgressMonitor monitor) {
-        ModelProvider modelProvider = build.getModelProvider();
-
-        Collection<EclipseProject> models = modelProvider.fetchModels(EclipseProject.class, fetchStrategy, tokenSource, monitor);
-        ImmutableSet.Builder<EclipseProject> result = ImmutableSet.builder();
-        for (EclipseProject model : models) {
-            result.addAll(HierarchicalElementUtils.getAll(model));
-        }
-        return result.build();
+    public static Collection<EclipseProject> fetchAllEclipseProjectsWithSyncTask(InternalGradleBuild build, CancellationTokenSource tokenSource, FetchStrategy fetchStrategy, IProgressMonitor monitor) throws Exception {
+        return build.withConnection(connection -> {
+            BuildEnvironment buildEnvironment = connection.getModel(BuildEnvironment.class);
+            GradleVersion gradleVersion = GradleVersion.version(buildEnvironment.getGradle().getGradleVersion());
+
+            if (supportsTaskExecution(gradleVersion)) {
+                return runTasksAndQueryCompositeEclipseModel(connection);
+            } else if (supportsCompositeBuilds(gradleVersion)) {
+                return queryCompositeEclipseModel(connection);
+            } else {
+                return queryEclipseModel(connection);
+            }
+        }, monitor);
+
+    }
+
+    private static boolean supportsTaskExecution(GradleVersion gradleVersion) {
```

Comment:
⭕️ You mean supportsPhasedActions?

Reply:
No, phased actions and task execution were introduced in Gradle 4.8, and 5.4 respectively. For Gradle <=4.8 <5.4 versions the model builder is not available for the `RunEclipseSynchronizationTasks` model.

Reply:
I just want to avoid confusion with "being able to execute tasks", which the TAPI could do from day1.

Reply:
Done.